File: node/src/web/statement.ts
import { desc, eq, sql } from "drizzle-orm";
import { Request, Response } from "express";
import { DB, clientAlias, transactionAlias } from "../db";

export const buildStatement = ({ db }: { db: DB }) => {
	const clientQuery = db
		.select({
			balance: clientAlias.balance,
			limit: clientAlias.limit,
			value: transactionAlias.value,
			description: transactionAlias.description,
			createdAt: transactionAlias.createdAt,
		})
		.from(clientAlias)
		.leftJoin(transactionAlias, eq(transactionAlias.clientId, clientAlias.id))
		.where(eq(clientAlias.id, sql.placeholder("id")))
		.orderBy(desc(transactionAlias.id))
		.limit(10);;

	return async (req: Request<{ id: number }>, res: Response) => {
		try {
			const [clients] = await Promise.all([
				clientQuery.execute({ id: req.params.id }),
			]);

			if (clients.length === 0) {
				return res.status(404).json({ message: "Client not found" }).end();
			}

			const [client] = clients;
			const transactions  = clients.filter((c) => c.value !== null);

			res.json({
				saldo: {
					total: client.balance,
					data_extrato: new Date().toISOString(),
					limite: client.limit,
				},
				ultimas_transacoes: (transactions ?? []).map((t) => ({
					valor: Math.abs(t.value!),
					tipo: t.value! > 0 ? "c" : "d",
					descricao: t.description,
					realizada_em: t.createdAt!.toISOString(),
				})),
			});
		} catch (error) {
			console.error("Statement error: ", error);
			res
				.status(500)
				.json({
					message: "Internal server error",
					details: error.message,
					body: req.body,
					clientId: req.params.id,
				})
				.end();
		}
	};
};

File: golang/statement.go
package main

import (
	"time"

	"github.com/gofiber/fiber/v2"
	"gorm.io/gorm"
)

func buildExtratoHandler(db *gorm.DB) func(c *fiber.Ctx) error {
	return func(c *fiber.Ctx) error {
		c.Set("content-type", "application/json")

		rows, err := db.Table("clients c").
			Joins("LEFT JOIN transactions t ON c.id = t.client_id").
			Where("c.id = ?", c.Params("id")).
			Select("c.\"balance\", c.\"limit\", coalesce(t.value,0) as value, coalesce(t.description,'') as description, t.created_at").
			Order("t.id desc").
			Limit(10).
			Rows()

		if err != nil {
			println(`Error calling extrato query: ` + err.Error())
			return c.Status(500).JSON(internalError)
		}
		defer rows.Close()

		if rows != nil && rows.Next() {
			var balance int
			var limit int
			var value int
			var description string
			var createdAt time.Time

			transactions := []map[string]interface{}{}
			err := rows.Scan(&balance, &limit, &value, &description, &createdAt)
			if err != nil && !ignoreScanErrors {
				println("Error parsing rows(extrato): " + err.Error())
				return c.Status(500).JSON(internalError)
			}

			for {
				if value == 0 {
					break
				}

				transactions = append(transactions, map[string]interface{}{
					"valor":        abs(value),
					"tipo":         toType(value),
					"descricao":    description,
					"realizada_em": createdAt,
				})

				if !rows.Next() {
					break
				}
				err = rows.Scan(&balance, &limit, &value, &description, &createdAt)
				if err != nil && !ignoreScanErrors {
					println("Error parsing rows(extrato, second): " + err.Error())
					return c.Status(500).JSON(internalError)
				}
			}

			return c.Status(200).JSON(fiber.Map{
				"saldo": fiber.Map{
					"total":        balance,
					"limite":       limit,
					"data_extrato": time.Now(),
				},
				"ultimas_transacoes": transactions,
			})
		}
		return c.Status(404).JSON(notFound)
	}
}
